python/src/main/python/drivers/runspv.py

```diff
@@ -0,0 +1,944 @@
+            if adb_can_fail('shell test -f ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/STARTED').returncode != 0:
+                continue
+            status = 'TIMEOUT'
+
+        assert status == 'TIMEOUT'
+
+        # DONE file indicates app is done.
+        if adb_can_fail('shell test -f ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/DONE').returncode == 0:
+            status = 'SUCCESS'
+            break
+
+        # Otherwise, keep looping/waiting while the app is still running.
+        # Quote >/dev/null otherwise this fails on Windows hosts.
+        if adb_can_fail('shell "pidof ' + ANDROID_LEGACY_APP + ' > /dev/null"').returncode == 0:
+            continue
+
+        # The app has crashed. Check for DONE file one more time.
+        if adb_can_fail('shell test -f ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/DONE').returncode == 0:
+            status = 'SUCCESS'
+            break
+
+        # App has terminated and there is no DONE file; this definitely looks like a crash.
+        status = 'CRASH'
+        break
+
+    # Grab log:
+    with open(logfile, 'w', encoding='utf-8', errors='ignore') as f:
+        adb_check('logcat -d', stdout=f)
+
+    # retrieve all files
+    adb_check('pull ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + ' ' + output_dir)
+
+    # Check sanity:
+    if status == 'SUCCESS':
+        sanity_before = output_dir + '/sanity_before.png'
+        sanity_after = output_dir + '/sanity_after.png'
+        if os.path.isfile(sanity_before) and os.path.isfile(sanity_after):
+            if not filecmp.cmp(sanity_before, sanity_after, shallow=False):
+                status = 'SANITY_ERROR'
+
+    # Check nondet:
+    if status == 'SUCCESS':
+        ref_image = output_dir + '/image_0.png'
+        if os.path.isfile(ref_image):
+            # If reference image is here then report nondet if any image is different or missing.
+            for i in range(1, NUM_RENDER):
+                next_image = output_dir + '/image_{}.png'.format(i)
+                if not os.path.isfile(next_image):
+                    status = 'UNEXPECTED_ERROR'
+                    with open(LOGFILE_NAME, 'a') as f:
+                        f.write('\n Not all images were produced? Missing image: {}\n'.format(i))
+                elif not filecmp.cmp(ref_image, next_image, shallow=False):
+                    status = 'NONDET'
+                    shutil.copy(ref_image, 'nondet0.png')
+                    shutil.copy(next_image, 'nondet1.png')
+
+    with open(logfile, 'a') as f:
+        f.write('\nSTATUS ' + status + '\n')
+        if status == 'UNEXPECTED_ERROR':
+            f.write('\n App did not start?\n')
+
+    with open(statusfile, 'w') as f:
+        f.write(status)
+
+    if status != 'SUCCESS':
+        # Something went wrong. Make sure we stop the app.
+        adb_can_fail('shell am force-stop ' + ANDROID_LEGACY_APP)
+
+    # Grab image if present.
+    image_path = ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/image_0.png'
```

Comment:
The file has already be retrieved on line 319, so we may drop this image file grabbing.

Reply:
Nice. I removed this.